molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h

```diff
@@ -0,0 +1,225 @@
+ *
+ * \see https://biopython.org/wiki/The_Biopython_Structural_Bioinformatics_FAQ
+ */
+struct IVW_MODULE_MOLVISBASE_API Chain {
+    size_t id;
+    std::string name;
+};
+
+/**
+ * represents a bond between two atoms and refers to the atom indices
+ */
+using Bond = std::pair<size_t, size_t>;
+
+/**
+ * \brief represents molecular data consisting of atoms and optional residue and chain information
+ * as well as bonds
+ *
+ * A MolecularData object requires the atoms to provide at least positions to be considered a valid
+ * molecular structure. Residue and chain information is optional. Information on bonds between
+ * atoms is also optional.
+ *
+ * Molecular data can consists of multiple chains. Chains consist of one or several residues which
+ * in turn contain individual atoms. This data structure assumes that only atoms refer to residues
+ * and chains via IDs. The hierarchical information is extracted in and used by MolecularStructure.
+ *
+ * \see MolecularStructure
+ */
+struct IVW_MODULE_MOLVISBASE_API MolecularData {
+    std::optional<std::string> source;
+
+    Atoms atoms;
+    std::vector<Residue> residues;
+    std::vector<Chain> chains;
+    std::vector<Bond> bonds;
+};
+
+/// residue's ID and chain ID for unique identification
+using ResidueID = std::pair<size_t, size_t>;
+
+}  // namespace molvis
+}  // namespace inviwo
+
+namespace std {
+template <>
+struct hash<inviwo::molvis::ResidueID> {
+    size_t operator()(const inviwo::molvis::ResidueID& resId) const {
+        size_t h = 0;
+        inviwo::util::hash_combine(h, resId.first);
+        inviwo::util::hash_combine(h, resId.second);
+        return h;
+    }
+};
+}  // namespace std
+
+namespace inviwo {
+namespace molvis {
+
+/**
+ * \brief a data structure holding molecular data and its acceleration structures
+ *
+ * A MolecularStructure holds and provides access to the hierarchical information of molecular data,
+ * i.e. chains, residues, and atoms, if existing. Acceleration structures are used for faster
+ * lookups of, e.g. querying all atoms in a particular residue.
+ *
+ * Note: acceleration structures can only be created if the molecular data provides information on
+ * residues or both residues and chains.
+ *
+ * \see MolecularData
+ */
+class IVW_MODULE_MOLVISBASE_API MolecularStructure {
+public:
+    /**
+     * \brief create a MolecularStructure from \p data
+     *
+     * Performs some basic sanity checks like ID consistency, equal size of atom attributes
+     * (positions, IDs, atomic numbers, etc.) and builds acceleration structures.
+     *
+     * @throws Exception if atoms refer to invalid residues/chains, an atom's chain ID mismatches
+     *         the chain ID of the residue, or if attributes stored in atom have different sizes
+     *         (empty attributes are ignored).
+     */
+    MolecularStructure(MolecularData data);
+    MolecularStructure() = delete;
+    ~MolecularStructure() = default;
+
+    const MolecularData& get() const;
```

Comment:
when reading the code below that used get(), I was confused by what it did. Think data() would be better,
might also consider adding pass through atoms(), chains() residues() and bonds().

Reply:
good point